rep-2004.rst

```diff
@@ -0,0 +1,527 @@
+Among those reasons listed is "To satisfy stakeholders", which is the main goal of requiring a code coverage policy for these high quality packages.
+It is summarized nicely:
+
+    For many projects, there are various actors who have an interest in software quality who may not be involved in the day-to-day development of the software (managers, technical leads, etc.)
+    Saying "we're going to write all the tests we really need" is not convincing:
+    They either need to trust entirely, or verify with ongoing close oversight (assuming they even have the technical understanding to do so.)
+    Providing measurable standards and explaining how they reasonably approximate actual goals is better.
+
+The other two reasons "To normalize team behavior" and "To keep yourself honest" are nice reasons to have code coverage goals, but are out of scope for this document.
+
+The general recommendation is to have at least line coverage and aim to achieve and maintain a high percentage of coverage (e.g. above 90%).
+This at least gives you and your stakeholders some confidence that all feature have basic tests.
+Any assurances beyond that would require branch coverage statistics and independent investigation of the tests and how they test the code.
+
+Performance Testing
+"""""""""""""""""""
+
+There are some cases where performance testing does not make sense to have.
+For example, it may be a good idea to have performance tests for a code generator (like ``rosidl_generator_cpp``), but it is not strictly required since its performance does not affect a runtime production system, and so in that case the package could claim to be 'Level 1' without performance tests if properly justified in the "quality declaration".
+
+However, if performance is a reasonable concern for use in a production system, then there must be performance tests and they should be used in conjunction with a regression policy which aims to prevent new versions of the package to be considerably slower without cause.
+Note, the performance regression policy should not prevent regressions, but instead should aim to detect them and either address them directly, plan to address them in the future, or when unavoidable (e.g. fixing a bug required more resources to be safe) explain why the regression has occurred in the memorandum of the change request that introduced it.
+
+Dependencies
+""""""""""""
+
+Each package should examine their direct runtime dependencies for their quality levels.
+Packages should not claim a quality level higher than their dependencies, unless it can be reasonably explained why they do not affect the quality of the package in question.
+
+An example of this would be build or "build tool" dependencies, which are only used during build time and do not impact the runtime quality of the package.
```

Comment:
Is this a statement we can make? It's not uncommon for quality standards to require the use of specific versions of build tools, compilers and related infrastructure.

```suggestion
An example of this would be build or "build tool" dependencies, which are only used during build time and do not affect the runtime quality of the package.
```

Reply:
I think this is a grey area. If we were certifying this for safety, then yes we'd look at everything that does or affects codegen (build system, compiler, etc.). But in this case I think the burden is too high.

There's an implicit assumption that a package must be built as documented (the documentation may be, built as a ROS 2 package, i.e. built with colcon on one of the tier 1 platforms based on the ROS 2 installation instructions and REP-2000), and if it is then the maintainers of the package have ensured the build does not adversely affect the quality of the package, regardless of the quality of the build tools (i.e. colcon doesn't need to be level 1 quality to produce binaries for level 1 quality packages). Again this isn't strong enough for safety certification perhaps, but I think it's good enough here and a trade-off worth making to keep the burden of proof lighter for package maintainers.

If you're referring to runtime things, like is the runtime or vm or standard library for my programing language good enough, I think this is also a grey area, but the general assumption is that this level of infrastructure (the language level) at least meet these criteria of quality here. If a user suggested otherwise, and the package maintainers were inclined to convince them otherwise, they could justify their dependency on that language in their package's "quality declaration". These dependencies are sometimes implicit at the package.xml level, i.e. we don't exec_depend on python3 or libstdc++ or anything.